src/Microsoft.Data.Analysis/DataFrameRowCollection.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Data.Analysis
+{
+
+    /// <summary>
+    /// Represents the rows of a <see cref="DataFrame"/>
+    /// </summary>
+    public class DataFrameRowCollection : IEnumerable<DataFrameRow>
+    {
+        private long _currentIndex = 0;
+        private DataFrame _dataFrame;
+
+        /// <summary>
+        /// Initializes a <see cref="DataFrameRowCollection"/>.
+        /// </summary>
+        internal DataFrameRowCollection(DataFrame dataFrame)
+        {
+            _dataFrame = dataFrame;
+        }
+
+        /// <summary>
+        /// An indexer to return the <see cref="DataFrameRow"/> at <paramref name="index"/>
+        /// </summary>
+        /// <param name="index">The row index</param>
+        public DataFrameRow this[long index]
+        {
+            get
+            {
+                List<object> row = GetRow(index);
+                return new DataFrameRow(row);
+            }
+        }
+
+        internal List<object> GetRow(long rowIndex)
+        {
+            var ret = new List<object>();
+            var columns = _dataFrame.Columns;
+            foreach (var column in columns)
+            {
+                ret.Add(column[rowIndex]);
+            }
+            return ret;
+        }
+
+        /// <summary>
+        /// Returns an enumerator of <see cref="DataFrameRow"/> objects
+        /// </summary>
+        public IEnumerator<DataFrameRow> GetEnumerator()
+        {
+            DataFrameRow _currentRow = new DataFrameRow(null);
+            while (_currentIndex < _dataFrame.RowCount)
+            {
+                _currentRow._row = GetRow(_currentIndex);
```

Comment:
You are mutating the rerference of `_row` here, so `_currentRow` (which gets yielded) will be updated in each iteration. (aside: `_currentRow` as local should be named `currentRow`)

To mitigate this
```c#
public IEnumerator<DataFrameRow> GetEnumerator()
{
    for (long currentIndex = 0; currentIndex < _dataFrame.RowCount; ++currentIndex)
    {
        yield return new DataFrameRow(GetRow(currentIndex));
    }
}
```